### PR TITLE
[logging] add file logging for simulation platform

### PR DESCRIPTION
**Pull request by Irving-cl**
This PR implements file logging for simulation (including 
VIRTUAL_TIME) as mentioned in #6471. With the PR, we
can add OT logs for failed cases on github actions.

This PR does the following things:
* Implements a file logging module under simulation platform.
* Add tests for the file logging module.
* Add a new LOG_OUTPUT `FILE` for logging to a file.
* Allow the log file name specfied by arguments. Also enhances cmd
line arguments parsing by using `getopt`. Currently our args parsing
is not robust and not flexible.

### CODE REVIEW

File: include/openthread/platform/logging.h [hunk 1]
```diff
@@ -106,6 +106,14 @@ extern "C" {
  */
 #define OT_LOG_LEVEL_DEBG 5
 
+/**
+ * Maximum length for log file.
+ *
+ * @note This is only used when using OPENTHREAD_CONFIG_LOG_OUTPUT_FILE.
+ *
+ */
+#define OT_LOG_FILE_NAME_MAX_LEN 128
```
Comment (abtink) [inline include/openthread/platform/logging.h:115]: This seems to be more platform implementation specific (I don't think OT core would need to be aware of file name size). How about removing it from `logging.h` and adding it under simul platform and `file_logging`?

**Review by abtink — APPROVED** [review]
Looks great. Thanks @Irving-cl . 👍 
Few smaller suggestions/questions below.

File: examples/platforms/simulation/file_logging.c [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    while (aPos < aLen)
+    {
+        uint16_t aBytesWrite = OT_MIN(aLen - aPos, LOG_BUFFER_SIZE - sLogBufferPos);
+        memcpy(sLogBuffer + sLogBufferPos, aLogString + aPos, aBytesWrite);
+
+        aPos += aBytesWrite;
+        sLogBufferPos += aBytesWrite;
+
+        if (sLogBufferPos == LOG_BUFFER_SIZE)
```
Comment (zhanglongxia) [inline examples/platforms/simulation/file_logging.c:93]: When the program crashed. Will the log be flashed to the log file?

File: examples/platforms/simulation/logging.c [hunk 3]
```diff
@@ -76,4 +77,31 @@ OT_TOOL_WEAK void otPlatLog(otLogLevel aLogLevel, otLogRegion aLogRegion, const
...
+
+    char         logString[512];
+    unsigned int offset;
+    int          charsWritten;
+    va_list      args;
+
+    offset = 0;
+
+    va_start(args, aFormat);
+    charsWritten = vsnprintf(&logString[offset], sizeof(logString) - offset, aFormat, args);
```
Comment (zhanglongxia) [inline examples/platforms/simulation/logging.c:94]: Here we copy the log to the `logString`, then we  copy the log to the `sLogBuffer` in the file `file_logging.c`. Can we avoid coping twice?

**Comment by simonlingoogle** [comment]
`argument parsing by getopt` and file logging are two separate enhancements? Should we use two separate PR for them?
@Irving-cl

**Review by simonlingoogle — COMMENTED** [review]
Can we use `stderr` redirection to a file for the same purpose? What is the advantage of adding a new LOG_OUTPUT type?

File: examples/platforms/simulation/logging.c [hunk 4]
```diff
@@ -76,4 +77,31 @@ OT_TOOL_WEAK void otPlatLog(otLogLevel aLogLevel, otLogRegion aLogRegion, const
...
+    OT_UNUSED_VARIABLE(aLogRegion);
+
+    char         logString[512];
+    unsigned int offset;
+    int          charsWritten;
+    va_list      args;
+
+    offset = 0;
+
+    va_start(args, aFormat);
```
Comment (bukepo) [inline examples/platforms/simulation/logging.c:93]: can we simply implement this function by:
```cpp
fprintf(stderr, aFormat, args);
```

File: examples/platforms/simulation/system.c [hunk 5]
```diff
@@ -65,54 +68,97 @@ static void handleSignal(int aSignal)
...
+
+        switch (c)
+        {
+        case 's':
+            gRadioCaps |= OT_RADIO_CAPS_SLEEP_TO_TX;
+            break;
+        case 't':
+            speedUpFactor = (uint32_t)strtol(optarg, NULL, 10);
+            break;
+#if OPENTHREAD_CONFIG_LOG_OUTPUT == OPENTHREAD_CONFIG_LOG_OUTPUT_FILE
```
Comment (bukepo) [inline examples/platforms/simulation/system.c:109]: I seems we can simply achieve the effect by using `stderr`.
```suggestion
#if OPENTHREAD_CONFIG_LOG_OUTPUT == OPENTHREAD_CONFIG_LOG_OUTPUT_STDERR
```

usage will be:
```
ot-cli-ftd 1 2> /path/to/logging_file
```

**Comment by Irving-cl** [comment]
Thanks for all the comments. After re-thinking on this, I agree we can log to file simply by using re-direction. I'll close this PR and add the `get_opt` enhancement in another PR.

**Comment by Irving-cl** [comment]
Let's still use this PR to add STDERR OUTPUT.

**Comment by simonlingoogle** [comment]
> Let's still use this PR to add STDERR OUTPUT.

Shouldn't stderr be the default output? @Irving-cl @bukepo

**Comment by simonlingoogle** [comment]
> Let's still use this PR to add STDERR OUTPUT.

Do we still needs LOG_OUTPUT=FILE?

**Comment by Irving-cl** [comment]
obselete

**Pull request closed without merging**